packages/core/src/search/details.ts

```diff
@@ -55,54 +72,110 @@ function setSearchParameterDetails(resourceType: string, code: string, details:
 function buildSearchParameterDetails(resourceType: string, searchParam: SearchParameter): SearchParameterDetails {
   const code = searchParam.code as string;
   const columnName = convertCodeToColumnName(code);
-  const expression = getExpressionForResourceType(resourceType, searchParam.expression as string)?.split('.');
-  if (!expression) {
-    // This happens on compound types
-    // In the future, explore returning multiple column definitions
-    return { columnName, type: SearchParameterType.TEXT };
-  }
-
-  let baseType = resourceType;
-  let elementDefinition = undefined;
-  let propertyType = undefined;
-  let array = false;
-
-  for (let i = 1; i < expression.length; i++) {
-    let propertyName = expression[i];
-    let hasArrayIndex = false;
-
-    const arrayIndexMatch = /\[\d+\]$/.exec(propertyName);
-    if (arrayIndexMatch) {
-      propertyName = propertyName.substring(0, propertyName.length - arrayIndexMatch[0].length);
-      hasArrayIndex = true;
-    }
+  const expressions = getExpressionsForResourceType(resourceType, searchParam.expression as string);
 
-    elementDefinition = getElementDefinition(baseType, propertyName);
-    if (!elementDefinition) {
-      throw new Error(`Element definition not found for ${resourceType} ${searchParam.code}`);
-    }
+  const builder: SearchParameterDetailsBuilder = {
+    elementDefinitions: [],
+    propertyTypes: new Set(),
+    array: false,
+  };
 
-    if (elementDefinition.max !== '0' && elementDefinition.max !== '1' && !hasArrayIndex) {
-      array = true;
+  for (const expression of expressions) {
+    const atomArray = flattenAtom(expression);
+    if (atomArray.length === 1 && atomArray[0] instanceof BooleanInfixOperatorAtom) {
+      builder.propertyTypes.add('boolean');
+    } else {
+      crawlSearchParameterDetails(builder, flattenAtom(expression), resourceType, 1);
     }
+  }
+
+  const result: SearchParameterDetails = {
+    columnName,
+    type: getSearchParameterType(searchParam, builder.propertyTypes),
+    elementDefinitions: builder.elementDefinitions,
+    array: builder.array,
+  };
+  setSearchParameterDetails(resourceType, code, result);
+  return result;
+}
+
+function crawlSearchParameterDetails(
+  details: SearchParameterDetailsBuilder,
+  atoms: Atom[],
+  baseType: string,
+  index: number
+): void {
+  const currAtom = atoms[index];
+
+  if (currAtom instanceof AsAtom) {
+    details.propertyTypes.add(currAtom.right.toString());
+    return;
+  }
 
+  if (currAtom instanceof FunctionAtom) {
+    handleFunctionAtom(details, currAtom);
+    return;
+  }
+
+  const propertyName = currAtom.toString();
+  let hasArrayIndex = false;
+  let nextIndex = index + 1;
+
+  if (nextIndex < atoms.length && atoms[nextIndex] instanceof IndexerAtom) {
+    hasArrayIndex = true;
+    nextIndex++;
+  }
+
+  const elementDefinition = getElementDefinition(baseType, propertyName);
```

Comment:
Should this use the new `InternalTypeSchema` instead?  I'm also happy to leave it this way for now and convert it over later

Reply:
Good question.  This file is already using `getElementDefinition`, so I'm inclined to wait until we're migrating everything after the validation cutover.  Good suggestion though